fili-core/src/main/java/com/yahoo/bard/webservice/table/availability/MetricUnionAvailability.java

```diff
@@ -0,0 +1,190 @@
+package com.yahoo.bard.webservice.table.availability;
+
+import com.google.common.collect.Sets;
+import com.yahoo.bard.webservice.data.config.names.TableName;
+import com.yahoo.bard.webservice.data.metric.MetricColumn;
+import com.yahoo.bard.webservice.table.Column;
+import com.yahoo.bard.webservice.table.PhysicalTable;
+import com.yahoo.bard.webservice.table.resolver.DataSourceConstraint;
+import com.yahoo.bard.webservice.util.IntervalUtils;
+import com.yahoo.bard.webservice.util.SimplifiedIntervalList;
+
+import org.joda.time.Interval;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import javax.validation.constraints.NotNull;
+
+/**
+ * An implementation of availability which which joins physical tables with the same dimension schema
+ * but non-intersecting metric schemas. Metrics on the <tt>MetricUnionAvailability</tt> are sourced from
+ * exactly one of the participating tables.
+ */
+public class MetricUnionAvailability implements Availability {
+
+    private final Set<TableName> dataSourceNames;
+    private final Set<MetricColumn> metricColumns;
+    private final Map<Availability, Set<MetricColumn>> metricAvailability;
+
+    /**
+     * Constructor.
+     *
+     * @param physicalTables  A set of <tt>PhysicalTable</tt>'s
+     * @param columns  The columns associated with all tables
+     */
+    public MetricUnionAvailability(@NotNull Set<PhysicalTable> physicalTables, @NotNull Set<Column> columns) {
+        metricColumns =  columns.stream()
+                .filter(column -> column instanceof MetricColumn)
+                .map(column -> (MetricColumn) column)
+                .collect(Collectors.toSet());
+
+        metricAvailability = physicalTables.stream()
+                .collect(
+                        Collectors.toMap(
+                                PhysicalTable::getAvailability,
+                                physicalTable ->
+                                        Sets.intersection(
+                                                physicalTable.getSchema().getColumns(MetricColumn.class),
+                                                metricColumns
+                                        )
+                        )
+                );
+
+        // validate metric uniqueness such that
+        // each table's underlying datasource schema don't have repeated metric column
+        Map<MetricColumn, Set<Availability>> duplicates = getDuplicateValue(metricAvailability);
+        if (!duplicates.isEmpty()) {
+            throw new IllegalArgumentException(
+                    String.format(
+                            "Metric columns are not unique - {}",
+                            duplicates
+                    )
+            );
+        }
+
+        dataSourceNames = physicalTables.stream()
```

Comment:
Good catch.
This is collecting physical table dictionary names and not datasource names.
We need to call 'getDataSourceNames' on the the underlying availabilities.